Feature login

IlanSirisky opened the pull request:
Login Feature with aws cognito and authentication middleware

IlanSirisky committed:
auth login route

IlanSirisky committed:
Merge branch 'dev' into Feature-Login

IlanSirisky committed:
Register endpoint

IlanSirisky committed:
confirm user endpoint

IlanSirisky committed:
Added auth token validation middleware

IlanSirisky committed:
Refactored DB Tables

IlanSirisky committed:
Refactored User Schema

IlanSirisky committed:
Adding user to db on registration

Starting with random pokemon

IlanSirisky committed:
Added count pokemon type counts endpoint

IlanSirisky committed:
removed console logs

Aslanchik-m commented on src/handlers/authHandler.ts:
```diff
@@ -0,0 +1,107 @@
+import userModel from "../models/userModel";
+import { getRandomPokemonId } from "../utils/randomStartingPokemon";
+
+dotenv.config();
+
+const userPool = new CognitoUserPool({
+  UserPoolId: process.env.COGNITO_USER_POOL_ID || "",
+  ClientId: process.env.COGNITO_CLIENT_ID || "",
+});
```
I'd recommend creating an `envs.ts` file in your root to contain all of your environment variables. At the moment you only have two or three, but as the application grows you will have many more. 
Another critical thing to note here is that - without these envs you cannot handle logins into your system, this means that users would not be able to connect.. lets think of a mechanism that will notify you (or any other developer that is debugging this system) that these envs are missing, could be a life saver in a production environment.

Aslanchik-m commented on src/handlers/pokemonHandler.ts:
```diff
@@ -52,9 +52,28 @@ const findPokemons = async (query: {
+const getPokemonTypesCount = async (userSub: string) => {
+  const result = await pokemonModel.getPokemonTypes(userSub);
+
+  const typesCount: { [key: string]: number } = {};
+
+  result.forEach((userPokemon) => {
+    userPokemon.Pokemon.profile?.types.forEach((type) => {
+      if (typesCount[type]) {
+        typesCount[type]++;
+      } else {
+        typesCount[type] = 1;
+      }
+    });
+  });
```
This screams `reduce` to me, try using reduce to gain the same functionality - it is a great method that would help you a lot in situations like this.

Aslanchik-m commented on src/middleware/authMiddleware.ts:
```diff
@@ -0,0 +1,59 @@
+AWS.config.update({ region: process.env.AWS_REGION });
+
+const cognito = new AWS.CognitoIdentityServiceProvider();
+const tokenCache = new NodeCache({ stdTTL: 600 }); // Cache tokens for 10 minutes
```
I'd also export this into an env - to be controlled from the outside

IlanSirisky committed:
Updated endpoints to query based on user_id

IlanSirisky committed:
Updated catch endpoint to add to user pokemon table

IlanSirisky committed:
Added envs.ts file

IlanSirisky committed:
Updated query to add isOwned column

Updated query to add isOwned column  based on usersPokemon table for indicator in frontend
